graspy/align/seedless_procrustes.py

```diff
@@ -0,0 +1,359 @@
+            raise ValueError(msg)
+        # check initial_Q argument
+        if initial_Q is not None:
+            if not isinstance(initial_Q, np.ndarray):
+                msg = f"Initial_Q must be np.ndarray or None, not {type(initial_Q)}"
+                raise TypeError(msg)
+            initial_Q = check_array(initial_Q, copy=True)
+            if initial_Q.shape[0] != initial_Q.shape[1]:
+                msg = "Initial_Q must be a square orthogonal matrix"
+                raise ValueError(msg)
+            if not np.allclose(initial_Q.T @ initial_Q, np.eye(initial_Q.shape[0])):
+                msg = "Initial_Q must be a square orthogonal matrix"
+                raise ValueError(msg)
+        # check initial_P argument
+        if initial_P is not None:
+            if not isinstance(initial_P, np.ndarray):
+                msg = f"Initial_P must be np.ndarray or None, not {type(initial_P)}"
+                raise TypeError(msg)
+            initial_P = check_array(initial_P, copy=True)
+            n, m = initial_P.shape
+            if not (
+                np.allclose(initial_P.sum(axis=0), np.ones(m) / m)
+                and np.allclose(initial_P.sum(axis=1), np.ones(n) / n)
+            ):
+                msg = (
+                    "Initial_P must be a doubly stochastic matrix "
+                    "(rows add up to (1/number of cols) "
+                    "and columns add up to (1/number of rows))"
+                )
+                raise ValueError(msg)
+
+        super().__init__()
+
+        self.optimal_transport_eps = optimal_transport_eps
+        self.optimal_transport_num_reps = optimal_transport_num_reps
+        self.optimal_transport_lambda = optimal_transport_lambda
+        self.iterative_num_reps = iterative_num_reps
+        self.init = init
+        self.initial_Q = initial_Q
+        self.initial_P = initial_P
+
+    def _optimal_transport(self, X, Y, Q):
+        # "E step" of the SeedlessProcrustes.
+        n, d = X.shape
+        m, _ = Y.shape
+        # initialize probability mass arrays & the cost matrix ; run sinkhorn
+        probability_mass_X = np.ones(n) / n
+        probability_mass_Y = np.ones(m) / m
+        cost_matrix = (
+            np.linalg.norm((X @ Q).reshape(n, 1, d) - Y.reshape(1, m, d), axis=2) ** 2
+        )
+        P = ot.sinkhorn(
+            a=probability_mass_X,
+            b=probability_mass_Y,
+            M=cost_matrix,
+            reg=self.optimal_transport_lambda,
+            numItermax=self.optimal_transport_eps,
+            stopThr=self.optimal_transport_eps,
+        )
+        return P
+
+    def _procrustes(self, X, Y, P):
+        # "M step" of the SeedlessProcurstes.
+        aligner = OrthogonalProcrustes()
+        Q = aligner.fit(X, P @ Y).Q_
+        return Q
+
+    def _iterative_ot(self, X, Y, Q):
+        # this P is not used. it is set to default in case numreps=0
+        P = np.ones((X.shape[0], Y.shape[0])) / (X.shape[0] * Y.shape[0])
```

Comment:
this confused me. are we trying to explicitly support numreps=0, and if so, why?

Reply:
yes. Youngser has been trying to use numreps=0 to triage good and bad starts in R code. when i tried it - it did not work well for me, but i decided to leave it as an option because "why not?". maybe there is something in it. shouldn't cause any issues, unless you explicitly set it to 0, in which case you probably know what you're doing.

Reply:
what is the actual algorithm that is run when numreps=0?

Reply:
for custom and sign-flips: initialize, compute objective, move on with life.

for "2d": initialize Q, initialize to this default P, compute objective. then choose best Q based on those objective values. set score and intial_Q to that.

Reply:
the idea here is the following: let's say d is big. 20. then you can probably run all inits for a couple iterations, but not for like a 100 of them. in my sims, convergence of this algo is quite fast in number of iterations, but is certainly more than a couple iterations.

Youngser has been trying to following: set k, X, k << X.

perform sign flips with k iterations. pick the best one over those. then run that for X. and he has been trying to go as low as k=0. i also gave it a shot. k=5 did okay, but k=0 did not work well for me, as i have mentioned previously, but i decided to leave this in as an option. (this triaging also highlights why keeping initial_Q is nice, because otherwise, you have to manually loop through all 2^d matrices and store which was the best one).

Reply:
feels a bit weird to me code-wise but i wont die on this hill if you want to keep this functionality